src/java/org/apache/cassandra/index/sai/disk/PostingList.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk;
+
+import java.io.Closeable;
+import java.io.IOException;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import org.apache.cassandra.utils.Throwables;
+
+/**
+ * Interface for advancing on and consuming a posting list.
+ */
+@NotThreadSafe
+public interface PostingList extends Closeable
+{
+    long OFFSET_NOT_FOUND = -1;
+    long END_OF_STREAM = Long.MAX_VALUE;
+
+    @Override
+    default void close() throws IOException {}
+
+    /**
+     * Retrieves the next segment row ID, not including row IDs that have been returned by {@link #advance(long)}.
+     *
+     * @return next segment row ID
+     */
+    long nextPosting() throws IOException;
+
+    long size();
+
+    /**
+     * Advances to the first row ID beyond the current that is greater than or equal to the
+     * target, and returns that row ID. Exhausts the iterator and returns {@link #END_OF_STREAM} if
+     * the target is greater than the highest row ID.
+     *
+     * Note: Callers must use the return value of this method before calling {@link #nextPosting()}, as calling
+     * that method will return the next posting, not the one to which we have just advanced.
+     *
+     * @param targetRowID target row ID to advance to
+     *
+     * @return first segment row ID which is >= the target row ID or {@link PostingList#END_OF_STREAM} if one does not exist
+     */
+    long advance(long targetRowID) throws IOException;
+
+    /**
+     * @return peekable wrapper of current posting list
+     */
+    default PeekablePostingList peekable()
```

Comment:
nit: I'm still thinking through this being at the interface level, and relying on a specific peekable implementation (even if it *is* the only one). Also, what happens if we call `peekable()` on `PeekablePostingList`? Maybe we should throw if that happens...seems like a clear programmer error...

Reply:
I agree with this. I have extracted the `PeekablePostingList` class out of this file and have added a `static makePeekable`. Instead of throwing if a `PeekablePostingList` is passed to this method, it just casts it to `PeekablePostingList` and returns it. Otherwise it does the normal wrap.